tools/export/sw4stm32/__init__.py

```diff
@@ -79,32 +81,103 @@ def __gen_dir(self, dirname):
     def __generate_uid(self):
         return "%0.9u" % randint(0, 999999999)
 
+    @staticmethod
+    def filter_dot(str):
+        """
+        This function removes ./ from str.
+        str must be converted with win_to_unix() before using this function.
+        """
+	if str == None:
+            return None
+        if str[:2] == './':
+            return str[2:]
+        return str
+
+    def build_excludelist(self):
+        self.source_folders = [self.filter_dot(s) for s in set(dirname(
+            src) for src in self.resources.c_sources + self.resources.cpp_sources + self.resources.s_sources)]
+        if '.' in self.source_folders:
+            self.source_folders.remove('.')
+        #print source_folders
+
+        top_folders = [f for f in set(s.split('/')[0] for s in self.source_folders)]
+        #print top_folders
+
+        for top_folder in top_folders:
+            #
+            for root, dirs, files in os.walk(top_folder, topdown=True):
+                # Paths returned by os.walk() must be split with os.dep
+                # to accomodate Windows weirdness.
+                parts = root.split(os.sep)
+                self.remove_unused('/'.join(parts))
+                
+    def remove_unused(self, path):
+	found = False
+        for used in self.include_path:
+            if path == used:
+                found = True
+	needToAdd = True
+        if not found:
+            for dir in self.exclude_dirs:
+               # Do not exclude subfolders from excluded folder
+               if path.find(dir+'/') != -1:
+                  needToAdd = False
+            if needToAdd:
+                self.exclude_dirs.append(path)
+
     def generate(self):
         fp_hardware = "no"
         fp_abi = "soft"
         core = self.toolchain.target.core
         if core == "Cortex-M4F" or core == "Cortex-M7F":
             fp_hardware = "fpv4-sp-d16"
-            fp_abi = "soft-fp"
+            fp_abi = "softfp"
         elif core == "Cortex-M7FD":
             fp_hardware = "fpv5-d16"
-            fp_abi = "soft-fp"
-            
+            fp_abi = "softfp"
+
+        self.resources.win_to_unix()
+
         libraries = []
         for lib in self.resources.libraries:
             l, _ = splitext(basename(lib))
             libraries.append(l[3:])
 
+        self.include_path = [self.filter_dot(s) for s in self.resources.inc_dirs]
+        print 'Include folders: {0}'.format(len(self.include_path))
```

Comment:
Python 3's print is not a keyword anymore and require parenthesis.
*Mbed uses Python 2, but if some day Python 3 is supported ([and it should be asap](https://pythonclock.org/)), best is to ease this work as much as possible.*